back/engines/commercial/smart_groups/app/lib/smart_groups/rules/email.rb

```diff
@@ -50,7 +50,18 @@ def initialize(predicate, value)
       self.value = value
     end
 
+    def cache_key_fragments
+      [User.all.cache_key_with_version, predicate, value]
```

Comment:
`cache_key_with_version` runs this query
```sql
SELECT COUNT(*) AS "size", MAX("users"."updated_at") AS timestamp FROM "users"
```
Though we don't have lots of records per table, this query makes the full scan and so can take as much time as the smart group query itself. Did you try checking how long it takes on production? (this [Metabase query](https://metabase.hq.citizenlab.co/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1RcbiAgXCJwcm9kdWN0aW9uX2NsMl9iYWNrX3Jhd1wiLlwiY2wyX3Jhd190ZW5hbnRzXCIuXCJob3N0XCIsXG4gIGNvdW50KFwicHJvZHVjdGlvbl9jbDJfYmFja19yYXdcIi5cImNsMl9yYXdfdXNlcnNcIi5cImlkXCIpXG5mcm9tXG4gIFwicHJvZHVjdGlvbl9jbDJfYmFja19yYXdcIi5cImNsMl9yYXdfdXNlcnNcIlxuam9pbiBcInByb2R1Y3Rpb25fY2wyX2JhY2tfcmF3XCIuXCJjbDJfcmF3X3RlbmFudHNcIiBvbiBcInByb2R1Y3Rpb25fY2wyX2JhY2tfcmF3XCIuXCJjbDJfcmF3X3RlbmFudHNcIi5cImlkXCIgPSBcInByb2R1Y3Rpb25fY2wyX2JhY2tfcmF3XCIuXCJjbDJfcmF3X3VzZXJzXCIuXCJ0ZW5hbnRfaWRcIlxuZ3JvdXAgYnkgXCJwcm9kdWN0aW9uX2NsMl9iYWNrX3Jhd1wiLlwiY2wyX3Jhd190ZW5hbnRzXCIuXCJob3N0XCJcbm9yZGVyIGJ5IGNvdW50KFwicHJvZHVjdGlvbl9jbDJfYmFja19yYXdcIi5cImNsMl9yYXdfdXNlcnNcIi5cImlkXCIpIGRlc2M7IiwidGVtcGxhdGUtdGFncyI6e319LCJkYXRhYmFzZSI6M30sImRpc3BsYXkiOiJ0YWJsZSIsInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==) can be useful)

If it's too long, maybe it makes sense to add an index on `updated_at`. Or, as an alternative, caching it with a static key, but just for 20 seconds.

Reply:
Not sure whether an index would even help with this. It's a valid concern indeed, could be worthwhile to run an explain or analyze to test it out.

Reply:
An index does not help here, I've attempted to add one.

PostgreSQL still scans the whole table, I assume this is because the select counts all rows anyway.

```
Aggregate  (cost=3.35..3.36 rows=1 width=8)
  ->  Seq Scan on users  (cost=0.00..3.28 rows=28 width=0)
```

Reply:
Maybe this trick could work https://dba.stackexchange.com/a/2071
Here's some explanation https://www.citusdata.com/blog/2016/10/12/count-performance/#distinct_counts_exact_index